### PR TITLE
Add usage to TextureViewDescriptor

### PR DESCRIPTION
**Pull request by sagudev**
**Description**
Spec says that we should obtain usage via TextureViewDescriptor: https://www.w3.org/TR/webgpu/#dom-gputextureviewdescriptor-usage, although by default it's is the same as in parent.

**Testing**
CTS run in servo: https://github.com/gfx-rs/wgpu/pull/6755#issuecomment-2547798268

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

**Comment by sagudev** [comment]
Given that hal::TextureUses are generated independently of any TextureUsage: https://github.com/gfx-rs/wgpu/blob/89a8867f9b3ae6dbbe8694890227a8ac0c9a6292/wgpu-core/src/device/resource.rs#L1251
I think we just need to do validation against desc.usage when doing stuff on TextureView.

File: wgpu-core/src/device/resource.rs
Comment (cwfitzgerald) [inline wgpu-core/src/device/resource.rs]: This should probably be `desc.usage == TextureUsages::FROM_PARENT`

File: wgpu-types/src/lib.rs
Comment (cwfitzgerald) [inline wgpu-types/src/lib.rs]: I'm torn here, on the one hand `= 0` bitflags behave really weird, but this is also a way to explicitly label the "none" use case. I think this is fine?

File: wgpu-core/src/resource.rs
Comment (cwfitzgerald) [inline wgpu-core/src/resource.rs]: The way we express this for other members is to make them optional, where None = inherit. I think we probably should do that here too.

File: wgpu-core/src/device/resource.rs
Comment (sagudev) [reply wgpu-core/src/device/resource.rs]: Both are same, but your seems more explicit so I will take that.

File: wgpu-core/src/resource.rs
Comment (sagudev) [reply wgpu-core/src/resource.rs]: Spec also works that way, all other members have null for inherit except this one which has 0, which is also the reason why I did it this way, so browser can do just `wgt::TextureUsages::from_bits()` without special handling of 0 case. I think that will also make it easier for wgpu-native (no experience there, so I might be wrong) and also for wgpu users coming from web where they would expect that wgt::TextureUsages(0) would be same as in web. This whole thing would be no brainer if bitflags would just support NonZeroU32.

Comment (cwfitzgerald) [reply wgpu-core/src/resource.rs]: I think in this case, the better api would be to be self-consistent, rather than consistent with the spec here - it's what users of a rusty api would come to expect and is self-explanatory for people coming from the JS spec.

**Comment by sagudev** [comment]
[CTS run in servo](https://github.com/sagudev/servo/actions/runs/12367767285/job/34519508975):
```
TIMEOUT [expected OK] /_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,createView:texture_view_usage:*

OK /_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,resource_usages,texture,in_render_misc:subresources,texture_view_usages:*
        PASS [expected FAIL] subtest: :bindingType="color-attachment";viewUsage=1
        PASS [expected FAIL] subtest: :bindingType="color-attachment";viewUsage=2
        PASS [expected FAIL] subtest: :bindingType="color-attachment";viewUsage=4
        PASS [expected FAIL] subtest: :bindingType="color-attachment";viewUsage=8
        PASS [expected FAIL] subtest: :bindingType="sampled-texture";viewUsage=1
        PASS [expected FAIL] subtest: :bindingType="sampled-texture";viewUsage=2
        PASS [expected FAIL] subtest: :bindingType="sampled-texture";viewUsage=8
        PASS [expected FAIL] subtest: :bindingType="sampled-texture";viewUsage=16
        PASS [expected FAIL] subtest: :bindingType="writeonly-storage-texture";viewUsage=1
        PASS [expected FAIL] subtest: :bindingType="writeonly-storage-texture";viewUsage=2
        And 2 more unexpected results...
```

**Comment by sagudev** [comment]
Servo is slow, but extending timeout and waiting half an hour gives:
1862/1862 PASS

File: wgpu-core/src/device/resource.rs
Comment (teoxoy) [inline wgpu-core/src/device/resource.rs:1341]: There is one more spot above that needs to be updated:

```rust
texture.hal_usage & mask_copy & mask_dimension & mask_mip_level
```

I think we can now remove `Texture.hal_usage` and can just call `conv::map_texture_usage(resolved_usage, aspects, format_flags)`.

File: wgpu-core/src/device/resource.rs
Comment (teoxoy) [inline wgpu-core/src/device/resource.rs:1101]: We don't seem to be checking that the format supports the usage:

image

**Review by teoxoy — CHANGES_REQUESTED** [review]

File: wgpu-core/src/device/resource.rs
Comment (sagudev) [reply wgpu-core/src/device/resource.rs:1341]: I did not feel confident doing that as there is a clear distinction between public usages (TextureUsage) and private usages (`TextureUses`).

File: wgpu-core/src/device/resource.rs
Comment (sagudev) [reply wgpu-core/src/device/resource.rs:1101]: There is some checking for render attachment later, but there is a lot more stuff missing/wrong ...

Anyway done in https://github.com/gfx-rs/wgpu/pull/6755/commits/c748f90a0944d8fc901b5764320045ce5f9375cb

File: wgpu-core/src/device/resource.rs
Comment (teoxoy) [inline wgpu-core/src/device/resource.rs]: We should be using `self.describe_format_features` since we might allow more usages on native with `TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES`.

File: wgpu-core/src/device/resource.rs
Comment (teoxoy) [reply wgpu-core/src/device/resource.rs:1341]: I don't think we will behave properly if we don't. Although the only case I know of that this matters in is: If you enable `bgra8unorm-storage` and create a texture with the `bgra8unorm` format & storage+renderable usage and then want to create a `bgra8unorm-srgb` renderable view from it (what we will pass to `wgpu-hal` is `bgra8unorm-srgb` with storage+renderable usage).

We can fix this later but I figured I'd point it out.

File: wgpu-core/src/device/resource.rs
Comment (sagudev) [reply wgpu-core/src/device/resource.rs]: The problem with that it also does check for required features which isn't per spec: https://github.com/gfx-rs/wgpu/blob/04a10405b3fc2c6cbab469af3d714e4209975378/wgpu-core/src/device/resource.rs#L3522

Comment (sagudev) [reply wgpu-core/src/device/resource.rs]: I am wrong, it is just part of content timeline: https://www.w3.org/TR/webgpu/#dom-gputexture-createview

**Review by cwfitzgerald — APPROVED** [review]
Looks good to me. Will let @teoxoy give final word before merging.

File: wgpu-core/src/device/resource.rs
Comment (sagudev) [reply wgpu-core/src/device/resource.rs]: Done in https://github.com/gfx-rs/wgpu/pull/6755/commits/9a9508b7c1b9647258e2d4ee2707e1bcbeff1945

**Review by teoxoy — APPROVED** [review]

**Pull request merged**
